lib/junk_drawer/rails/bulk_updatable.rb

```diff
@@ -11,14 +11,30 @@ def bulk_update(objects)
       objects = objects.select(&:changed?)
       return unless objects.any?
 
-      changed_attributes = extract_changed_attributes(objects)
-      query = build_query_for(objects, changed_attributes)
+      unique_objects = uniquify_and_merge(objects)
+      changed_attributes = extract_changed_attributes(unique_objects)
+      query = build_query_for(unique_objects, changed_attributes)
       connection.execute(query)
       objects.each(&:clear_changes_information)
     end
 
   private
 
+    def uniquify_and_merge(objects)
+      grouped_objects = objects.group_by(&:id).values
+      grouped_objects.each do |duplicates|
+        attrs = duplicates.each_with_object({}) do |duplicate, changes|
```

Comment:
Given that in the general case we only have one instance of a record, could we check for that at the top of the loop and simply return it (rather than running through its changes and resetting the `attributes`)?

Or filter `group_objects` to those with `values` length > 1 before entering the `.each` loop.

Not a huge deal, it just feels like this does extra work for every record when the relevant scenario might not be very frequent.

Reply:
I added a guard clause to skip when there is only 1 item in the group.